File: Player/src/main/java/ru/testsimpleapps/coloraudioplayer/managers/player/playlist/cursor/CursorFactory.java
package ru.testsimpleapps.coloraudioplayer.managers.player.playlist.cursor;

import ru.testsimpleapps.coloraudioplayer.App;
import ru.testsimpleapps.coloraudioplayer.managers.player.data.PlayerConfig;
import ru.testsimpleapps.coloraudioplayer.managers.player.playlist.IPlaylist;

public class CursorFactory {

    private static volatile CursorPlaylist sCursorPlaylist = null;
    private static volatile CursorPlaylist sCursorPlaylistForView = null;

    private CursorFactory() {}

    public static IPlaylist getInstance() {
        CursorPlaylist cursorPlaylist = sCursorPlaylist;
        if (sCursorPlaylist == null) {
            synchronized (CursorFactory.class) {
                cursorPlaylist = sCursorPlaylist;
                if (cursorPlaylist == null) {
                    sCursorPlaylist = cursorPlaylist = new CursorPlaylist(App.getContext());
                    sCursorPlaylist.setCursor(PlayerConfig.getInstance().getPlaylistId(), PlayerConfig.getInstance().getPlaylistSort(),
                            PlayerConfig.getInstance().getPlaylistSortOrder());
                    sCursorPlaylist.goToId(PlayerConfig.getInstance().getTrackId());
                }
            }
        }

        return cursorPlaylist;
    }

    public static IPlaylist newInstance() {
        final CursorPlaylist cursorPlaylist = (CursorPlaylist) getInstance();
        synchronized (CursorFactory.class) {
            cursorPlaylist.setCursor(PlayerConfig.getInstance().getPlaylistId(), PlayerConfig.getInstance().getPlaylistSort(),
                    PlayerConfig.getInstance().getPlaylistSortOrder());
            cursorPlaylist.goToId(PlayerConfig.getInstance().getTrackId());
        }

        return cursorPlaylist;
    }

    public static IPlaylist getCopyInstance() {
        final CursorPlaylist cursorPlaylist = sCursorPlaylist;
        CursorPlaylist viewCursorPlaylist = sCursorPlaylistForView;
        if (cursorPlaylist != null) {
            synchronized (CursorFactory.class) {
                if (viewCursorPlaylist != null) {
                    viewCursorPlaylist.closeCursor();
                }
                sCursorPlaylistForView = viewCursorPlaylist = (CursorPlaylist) cursorPlaylist.clone();
            }
        }

        return viewCursorPlaylist;
    }

    public static void close() {
        final CursorPlaylist cursorPlaylist = sCursorPlaylist;
        final CursorPlaylist viewCursorPlaylist = sCursorPlaylistForView;

        if (cursorPlaylist != null) {
            synchronized (CursorFactory.class) {
                cursorPlaylist.closeCursor();
                if (viewCursorPlaylist != null) {
                    viewCursorPlaylist.closeCursor();
                }
            }
        }
    }

}

File: player/src/main/java/ru/testsimpleapps/coloraudioplayer/managers/player/playlist/cursor/CursorFactory.kt
package ru.testsimpleapps.coloraudioplayer.managers.player.playlist.cursor

import ru.testsimpleapps.coloraudioplayer.app.App
import ru.testsimpleapps.coloraudioplayer.managers.player.data.PlayerConfig
import ru.testsimpleapps.coloraudioplayer.managers.player.playlist.IPlaylist
import ru.testsimpleapps.coloraudioplayer.service.PlayerService

object CursorFactory {

    @Volatile
    private var sCursorPlaylist: CursorPlaylist? = null
    @Volatile
    private var sCursorPlaylistForView: CursorPlaylist? = null

    val instance: IPlaylist
        get() {
            var cursorPlaylist = sCursorPlaylist
            if (sCursorPlaylist == null) {
                synchronized(CursorFactory::class.java) {
                    cursorPlaylist = sCursorPlaylist
                    if (cursorPlaylist == null) {
                        cursorPlaylist = CursorPlaylist(App.instance)
                        sCursorPlaylist = cursorPlaylist
                        sCursorPlaylist!!.setCursor(PlayerConfig.instance.playlistId, PlayerConfig.instance.playlistSort!!,
                                PlayerConfig.instance.playlistSortOrder!!)
                        sCursorPlaylist!!.goToId(PlayerConfig.instance.trackId)
                    }
                }
            }

            return cursorPlaylist!!
        }

    val copyInstance: IPlaylist?
        get() {
            val cursorPlaylist = sCursorPlaylist
            var viewCursorPlaylist = sCursorPlaylistForView
            if (cursorPlaylist != null) {
                synchronized(CursorFactory::class.java) {
                    if (viewCursorPlaylist != null) {
                        viewCursorPlaylist!!.closeCursor()
                    }
                    viewCursorPlaylist = cursorPlaylist.clone() as CursorPlaylist
                    sCursorPlaylistForView = viewCursorPlaylist
                }
            }

            return viewCursorPlaylist
        }

    fun newInstance(): IPlaylist {
        val cursorPlaylist = instance as CursorPlaylist
        synchronized(CursorFactory::class.java) {
            val playerConfig = PlayerConfig.instance
            cursorPlaylist.setCursor(playerConfig.playlistId, playerConfig.playlistSort!!, playerConfig.playlistSortOrder!!)
            cursorPlaylist.goToId(PlayerConfig.instance.trackId)
        }

        return cursorPlaylist
    }

    fun close() {
        val cursorPlaylist = sCursorPlaylist
        val viewCursorPlaylist = sCursorPlaylistForView

        if (cursorPlaylist != null) {
            synchronized(CursorFactory::class.java) {
                cursorPlaylist.closeCursor()
                viewCursorPlaylist?.closeCursor()
            }
        }
    }

}
